changed lcm and gcd implementations for when just one arg is infinite

Priynsh opened the pull request:
Fixes #56991 by adding conditions for both gcd and lcm cases. Also fixes tests for the same i.e. lot of tests regarding gcdx and lcm are removed. Two relevant tests have been added as well.

LilithHafner commented:
Thanks! Looking at this I'm realizing there is another way to define `lcm` and `gcd` that allows `lcm(1//1, 1//0)`, and even `lcm(0//1, 1//0)` to be well-defined.

[Currently](https://github.com/JuliaLang/julia/pull/56992), we define

> a is a multiple of b if there exists an integer m such that a=mb.

If we change that to

> a is a multiple of b if there exists an integer or 1//0 or false m such that a=mb.

Then the following are correct:

```julia-repl
julia> lcm(1//0, 1//1)
1//0 # currently 1//1

julia> lcm(1//0, 0//1)
0//1 # current behavior

julia> gcd(1//0, 1//1)
1//1 # currently 1//0

julia> gcd(1//0, 0//1)
1//0 # current behavior
```

Priynsh commented:
since the discussion was done, this is the new implementation. Minor change -

> a is a multiple of b if there exists an integer or 1//0 or -1//0 or false m such that a=mb.

LilithHafner commented:
Triage wants to add the definition of multiple to gcd to match lcm (but it should be the definition of divides). Triage does not like Lilith's idea of extending integers to include 1//0, -1//0, and false in this case.

Priynsh commented:
I’ve updated the implementation to throw an error when one of the arguments is infinite in gcd/lcm and fixed the related tests to reflect this change. Additionally, I’ve added a definition for 'multiple' in the docs :)

LilithHafner commented on base/intfuncs.jl:
```diff
@@ -96,6 +96,8 @@ end
 
 Least common (positive) multiple (or zero if any argument is zero).
 The arguments may be integer and rational numbers.
+``a`` is a multiple of ``b`` if there exists an integer ``m`` such
+that ``b*m=a``.
 
 ``a`` is a multiple of ``b`` if there exists an integer ``m`` such
 that ``a=mb``.
```
This note is already present; it previously looked like it wasn't because the base commit was out of date.

LilithHafner commented on base/rational.jl:
```diff
@@ -564,8 +564,18 @@ end
-gcd(x::Rational, y::Rational) = unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
-lcm(x::Rational, y::Rational) = unsafe_rational(lcm(x.num, y.num), gcd(x.den, y.den))
+function gcd(x::Rational, y::Rational)
+    if (isinf(x) && !isinf(y)) || (!isinf(x) && isinf(y))
```
I think this is more clear:
```suggestion
    isinf(x) != isinf(y)
```

LilithHafner commented on base/rational.jl:
```diff
@@ -564,8 +564,18 @@ end
-lcm(x::Rational, y::Rational) = unsafe_rational(lcm(x.num, y.num), gcd(x.den, y.den))
+function gcd(x::Rational, y::Rational)
+    if (isinf(x) && !isinf(y)) || (!isinf(x) && isinf(y))
+        throw(ArgumentError("lcm is not defined"))
```
```suggestion
        throw(ArgumentError("lcm is not defined between infinite and finite numbers"))
```

LilithHafner commented on base/rational.jl:
```diff
@@ -564,8 +564,18 @@ end
+    if (isinf(x) && !isinf(y)) || (!isinf(x) && isinf(y))
+        throw(ArgumentError("lcm is not defined"))
+    end
+    return unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
```
```suggestion
    unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
```
Most of this file uses implicit return stlye

LilithHafner commented on test/rational.jl:
```diff
@@ -649,59 +649,39 @@ end
+        @test_throws ArgumentError lcm(T(1)//T(0), T(1)//T(1))
     end
 end
-
```
Let's remove this unrelated whitespace change

LilithHafner commented on test/rational.jl:
```diff
@@ -649,59 +649,39 @@ end
 
         @test lcm([T(5), T(2), T(1)//T(2)]) === T(10)//T(1)
         @test lcm(T(5), T(2), T(1)//T(2)) === T(10)//T(1)
+
+        @test_throws ArgumentError gcd(T(1)//T(1), T(1)//T(0))
+        @test_throws ArgumentError gcd(T(1)//T(0), T(0)//T(1))
+
+        @test_throws ArgumentError lcm(T(1)//T(1), T(1)//T(0))
+        @test_throws ArgumentError lcm(T(1)//T(0), T(1)//T(1))
```
For tests, let's just change all the existing tests to @test_throws; that way we know we're not losing coverage. It will also make the blame/diff easier to review.

Priynsh replied:
done :)

LilithHafner reviewed (commented):
Thanks! I've left some inline comments and this looks good; we should be able to get it merged.

LilithHafner commented on test/rational.jl:
```diff
@@ -649,56 +649,62 @@ end
+
+        @test_throws ArgumentError gcd(T(1)//T(1), T(1)//T(0))
+        @test_throws ArgumentError gcd(T(1)//T(0), T(0)//T(1))
+
+        @test_throws ArgumentError lcm(T(1)//T(1), T(1)//T(0))
+        @test_throws ArgumentError lcm(T(1)//T(0), T(1)//T(1))
```
```suggestion
```
These tests are included above

LilithHafner commented on base/rational.jl:
```diff
@@ -564,8 +564,18 @@ end
+    unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
+end
+function lcm(x::Rational, y::Rational)
+    if (isinf(x) && !isinf(y)) || (!isinf(x) && isinf(y))
```
```suggestion
    isinf(x) != isinf(y)
```
Same comment as above

LilithHafner commented on base/rational.jl:
```diff
@@ -564,8 +564,18 @@ end
-gcd(x::Rational, y::Rational) = unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
-lcm(x::Rational, y::Rational) = unsafe_rational(lcm(x.num, y.num), gcd(x.den, y.den))
+function gcd(x::Rational, y::Rational)
+    isinf(x) != isinf(y)
+        throw(ArgumentError("lcm is not defined between infinite and finite numbers"))
+    end
+    unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
+end
+function lcm(x::Rational, y::Rational)
+    isinf(x) != isinf(y)
```
```suggestion
    if isinf(x) != isinf(y)
        throw(ArgumentError("lcm is not defined between infinite and finite numbers"))
    end
    unsafe_rational(gcd(x.num, y.num), lcm(x.den, y.den))
end
function lcm(x::Rational, y::Rational)
    if isinf(x) != isinf(y)
```
This is a typo, we're missing `if` here.

LilithHafner reviewed (approved):
Thanks for this PR!
